### PR TITLE
Support for DS3231 RTC clock

### PR DESCRIPTION
**Pull request by snakeye**
## Description:

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#959

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ds3231/__init__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import i2c
+from esphome.const import (CONF_ID)
+
+DEPENDENCIES = ['i2c']
```
Comment (jesserockz) [inline esphome/components/ds3231/__init__.py:6]: @snakeye You need to add this here and then run `script/build_codeowners.py`

```suggestion
CODEOWNERS = ['@snakeye']
DEPENDENCIES = ['i2c']
```

File: esphome/components/ds3231/sensor/__init__.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import i2c, sensor
+from esphome.const import CONF_ID, CONF_TEMPERATURE, ICON_THERMOMETER, UNIT_CELSIUS
+
+from .. import CONF_DS3231_ID, DS3231Component, ds3231_ns
+
+DEPENDENCIES = ['ds3231']
+
+RTCClock = ds3231_ns.class_('DS3231SensorComponent', cg.PollingComponent, i2c.I2CDevice)
```
Comment (jesserockz) [inline esphome/components/ds3231/sensor/__init__.py:10]: This should be renamed to match the C++ class name.

**Review by jesserockz — CHANGES_REQUESTED** [review]
Currently the parent component is required by the sensor and the time platforms, but it doesn't actually do anything because you have set the sensor and time platforms to be their own i2c devices...
Normally only the parent has the knowledge of i2c and the child platforms will either register themselves with the parent and get published to, or they will fetch an update by asking the parent for one.

Basically only one components code should be responsible for the actual communication with a device, being the parent component in the case of multi-platform components.

There would be too many changes to make suggestions in here sorry.

File: esphome/components/ds3231/sensor/__init__.py (hunk 2 above)
Comment (snakeye) [reply esphome/components/ds3231/sensor/__init__.py:10]: Thank you for the feedback!

Could you please point the component that is already working as described? I can use it as an example and update my code.

Comment (jesserockz) [reply esphome/components/ds3231/sensor/__init__.py:10]: Looking through the code, the simplest I found might be the `apds9960`, it doesn't have specific custom child platform  c++ code, but the parent registers the platforms then publishes their state.
Otherwise, it's not I2C, but the group of `tuya` components also do this.

Comment (badbadc0ffee) [reply esphome/components/ds3231/sensor/__init__.py:10]: Nice to see that support for more RTCs is on its way! Maybe we can have a quick chat: I like the idea to publish the temperature along with the time information. I am only wondering if we can refactor our drivers so they share some of their code. The time registers of the DS3231 and DS1307 are more or less identical in function, configuration registers have similar functions. One approach that I have in mind would be a generic I2C RTC implementation that accepts a template struct with the register definitions. This driver is currently lacking an interface to write the time back into the RTC - if I didn't get soemthing wrong; I'm reading this on my mobile phones tiny screen these days. ;-)

Comment (snakeye) [reply esphome/components/ds3231/sensor/__init__.py:10]: Yes, I think it makes sense to unify the code for DS1307/DS3231 - they have a lot in common. 

Setting the time works as follows:

```
time:
  - platform: smtp
     on_time:
       - seconds: /60
          minutes: /5
          then:
            lambda: |-
              auto time = id(sntp_time).now();
              if(time.is_valid()) {
                id(rtc_time).set_epoch_time(time.timestamp);
              }
```

Comment (nuttytree) [reply esphome/components/ds3231/sensor/__init__.py:10]: So I have need of this component, in particular I need the alarm functionality so I started working on this also not realizing someone else was.  I also started going down the route of just making this a time component instead of a parent DS3231 component with separate time and sensor components (partially because I hadn't planned on building the temp sensor).  After seeing this PR I agree the components should be separate and I am thinking the alarm function could stay with the parent DS3231 component as it really doesn't fit with time component.  I pushed what I have so far if you want to take a look [add-ds3231-component](https://github.com/nuttytree/esphome/tree/add-ds3231-component/esphome/components/ds3231).  I used structs/unions to represent the registers like the ds1307 component, I think it makes it clearer what you are reading/writing from/to.  I would be happy to help on this PR or if you want to complete what you have and I can add the alarm functionality afterwards.

Comment (snakeye) [reply esphome/components/ds3231/sensor/__init__.py:10]: I'm planning to refactor the component, but do not have enough time at the moment. Hopefully, I can get to it over the weekend.

Comment (nuttytree) [reply esphome/components/ds3231/sensor/__init__.py:10]: Sounds good let me know if I can do anything to help.
